anvil/src/eth/backend/mem/mod.rs

```diff
@@ -0,0 +1,1031 @@
+        self.db.read().basic(address).nonce.into()
+    }
+
+    /// Sets the coinbase address
+    pub fn set_coinbase(&self, address: Address) {
+        self.env.write().block.coinbase = address;
+    }
+
+    /// Sets the nonce of the given address
+    pub fn set_nonce(&self, address: Address, nonce: U256) {
+        self.db.write().set_nonce(address, nonce.try_into().unwrap_or(u64::MAX));
+    }
+
+    /// Sets the balance of the given address
+    pub fn set_balance(&self, address: Address, balance: U256) {
+        self.db.write().set_balance(address, balance);
+    }
+
+    /// Sets the code of the given address
+    pub fn set_code(&self, address: Address, code: Bytes) {
+        self.db.write().set_code(address, code);
+    }
+
+    /// Sets the value for the given slot of the given address
+    pub fn set_storage_at(&self, address: Address, slot: U256, val: U256) {
+        self.db.write().set_storage_at(address, slot, val);
+    }
+
+    pub fn gas_limit(&self) -> U256 {
+        self.env().read().block.gas_limit
+    }
+
+    /// Returns the current basefee
+    pub fn base_fee(&self) -> U256 {
+        self.fees.base_fee()
+    }
+
+    /// Sets the current basefee
+    pub fn set_base_fee(&self, basefee: U256) {
+        self.fees.set_base_fee(basefee)
+    }
+
+    /// Returns the current gas price
+    pub fn gas_price(&self) -> U256 {
+        self.fees.gas_price()
+    }
+
+    /// Sets the gas price
+    pub fn set_gas_price(&self, price: U256) {
+        self.fees.set_gas_price(price)
+    }
+
+    pub fn elasticity(&self) -> f64 {
+        self.fees.elasticity()
+    }
+
+    /// Validates the transaction's validity when it comes to nonce, payment
+    ///
+    /// This is intended to be checked before the transaction makes it into the pool and whether it
+    /// should rather be outright rejected if the sender has insufficient funds.
+    pub fn validate_transaction(
+        &self,
+        tx: &PendingTransaction,
+    ) -> Result<(), InvalidTransactionError> {
+        let sender = *tx.sender();
+        let tx = &tx.transaction;
+        let account = self.db.read().basic(sender);
+
+        // check nonce
+        if tx.nonce().as_u64() < account.nonce {
+            return Err(InvalidTransactionError::Payment)
+        }
+
+        let max_cost = tx.max_cost();
+        let value = tx.value();
+        // check sufficient funds: `gas * price + value`
+        let req_funds = max_cost.checked_add(value).ok_or(InvalidTransactionError::Payment)?;
+
+        if account.balance < req_funds {
+            return Err(InvalidTransactionError::Payment)
+        }
+        Ok(())
+    }
+
+    /// Mines a new block and stores it.
+    ///
+    /// this will execute all transaction in the order they come in and return all the markers they
+    /// provide.
+    ///
+    /// TODO(mattsse): currently we're assuming all transactions are valid:
+    ///  needs an additional validation step: gas limit, fee
```

Comment:
should we add this before merging?

Reply:
most of this was added, and all tx here are validated, but only at the point when they entered the mempool, but could be invalid at this stage (if the previous tx were executed in between) so there needs to be another validation step before including in the block

will add this